app/notification.go

```diff
@@ -458,7 +458,39 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
 		message.Add("mentions", model.ArrayToJson(mentionedUsersList))
 	}
 
-	a.Publish(message)
+	var previewedPostID string
+	if val, ok := post.GetProp(model.POST_PROPS_PREVIEWED_POST).(string); ok {
+		previewedPostID = val
+	}
+	var previewedChannel *model.Channel
+	if model.IsValidId(previewedPostID) {
+		previewedPost, err := a.GetSinglePost(previewedPostID)
+		if err != nil {
+			return nil, err
+		}
+
+		channelMembers, err := a.GetChannelMembersPage(channel.Id, 0, 10000000)
+		if err != nil {
+			return nil, err
+		}
+
+		previewedChannel, err = a.GetChannel(previewedPost.ChannelId)
+		if err != nil {
+			return nil, err
+		}
+
+		for _, cm := range *channelMembers {
+			postForIndividualUser := post.Clone()
+			if !a.HasPermissionToReadChannel(cm.UserId, previewedChannel) {
+				postForIndividualUser.Metadata.Embeds[0].Data = nil
+			}
+			message.Broadcast.UserId = cm.UserId
+			message.Add("post", postForIndividualUser.ToJson())
```

Comment:
Since these are mutating the message, do we need to copy the message in each of these loops? I'm thinking both because future iterations of the loop might want the non-cleared metadata, but also because we would potentially be getting some racy behaviour here assuming `a.Publish` does something asynchronous.

Reply:
Yeah, only has a small performance for some peace of mind.